media-proxy/src/libfabric_dev.c

```diff
@@ -90,42 +92,186 @@ static int rdma_init_fabric(libfabric_ctx *rdma_ctx, struct fi_info *hints)
+        ret = rdma_getaddrinfo((*ctx)->remote_ip, (*ctx)->remote_port, &rai_hints, &rai_res);
+        if (ret) {
+            printf("[rdma_init] ERROR: rdma_getaddrinfo (TX) returned %d\n", ret);
+            goto err;
+        }
+
+        // Handle source address
+        if (rai_res->ai_src_len > 0) {
+            hints->src_addr = malloc(rai_res->ai_src_len);
+            if (!hints->src_addr) {
+                ret = -ENOMEM;
+                goto err;
+            }
+            memcpy(hints->src_addr, rai_res->ai_src_addr, rai_res->ai_src_len);
+            hints->src_addrlen = rai_res->ai_src_len;
+        }
+
+        // Handle destination address
+        if (rai_res->ai_dst_len > 0) {
+            hints->dest_addr = malloc(rai_res->ai_dst_len);
+            if (!hints->dest_addr) {
+                ret = -ENOMEM;
+                goto err;
+            }
+            memcpy(hints->dest_addr, rai_res->ai_dst_addr, rai_res->ai_dst_len);
+            hints->dest_addrlen = rai_res->ai_dst_len;
+        }
+    }
+
+    /* Create fabric info, domain and fabric */
+    // For RX, pass hostname and port to enforce binding
+    if ((*ctx)->kind == KIND_RECEIVER) {
+        ret = fi_getinfo(FI_VERSION(2, 0), (*ctx)->local_ip, (*ctx)->local_port, 
+                        FI_SOURCE, hints, &(*ctx)->info);
+    } else {
+        // For TX, use the remote address info to force connectivity to specific endpoint
+        printf("[rdma_init] TX - Enforcing connection to remote %s:%s\n", (*ctx)->remote_ip,
+               (*ctx)->remote_port);
+
+        ret = fi_getinfo(FI_VERSION(2, 0), (*ctx)->remote_ip, (*ctx)->remote_port, 0, hints,
+                         &(*ctx)->info);
+    }
     if (ret) {
-        libfabric_dev_ops.rdma_deinit(ctx);
-        ERROR("Failed to initialize RDMA device");
-        return ret;
+        fprintf(stderr, "[rdma_init] ERROR: fi_getinfo returned %d (%s)\n", ret, fi_strerror(-ret));
+        goto err;
     }
 
+    ret = fi_fabric((*ctx)->info->fabric_attr, &(*ctx)->fabric, NULL);
+    if (ret) {
+        printf("[rdma_init] ERROR: fi_fabric returned %d\n", ret);
+        goto err_info;
+    }
+
+    ret = fi_domain((*ctx)->fabric, (*ctx)->info, &(*ctx)->domain, NULL);
+    if (ret) {
+        printf("[rdma_init] ERROR: fi_domain returned %d\n", ret);
+        goto err_fabric;
+    }
+
+    /* Clean up temporary resources */
+    rdma_freeaddrinfo(rai_res);
+    rdma_freehints(hints);  // We've already stored the info in ctx->info
+
+    (*ctx)->is_initialized = true;
+    printf("[rdma_init] Device context successfully initialized\n");
     return 0;
+
+err_fabric:
+    fi_close(&(*ctx)->fabric->fid);
+    (*ctx)->fabric = NULL;
+    
+err_info:
+    fi_freeinfo((*ctx)->info);
+    (*ctx)->info = NULL;
+    free(*ctx);
+    *ctx = NULL;
+    
+err:
+    if (rai_res) {
+        rdma_freeaddrinfo(rai_res);
+    }
+    if (hints) {
+        rdma_freehints(hints);
+    }
```

Comment:
It is safe to avoid NULL checking here
```suggestion
    rdma_freeaddrinfo(rai_res);
    rdma_freehints(hints);
```